app/services/external_apis/ror_service.rb

```
@@ -158,53 +159,98 @@ def parse_results(json:)
+      #     {"lang": "en", "types": ["ror_display","label"], "value": "Harvard University"},
+      #     {"lang": "es","types": ["label"], "value": "Universidad de Harvard"}
+      # ]
+      def sort_name(item)
+        item['names']
+          &.find { |n| n['types']&.include?('ror_display') }
+          &.fetch('value', nil)
+      end
+
+      # Returns the website link value
+      # "links": [
+      #   { "type": "website", "value": "https://example.edu" },
+      #   { "type": "Wikipedia", "value": "https://en.wikipedia.org/wiki/Example_University" }
+      # ]
+      def org_url(item)
+        links = item['links']
+        # links must be an array
+        return nil unless links.is_a?(Array)
+
+        links.find { |l| l['type'] == 'website' }&.fetch('value', nil)
+      end
+
       # Org names are not unique, so include the Org URL if available or
       # the country. For example:
       #    "Example College (example.edu)"
       #    "Example College (Brazil)"
       def org_name(item:)
-        return '' unless item.present? && item['name'].present?
+        name = sort_name(item)
 
         country = item.fetch('country', {}).fetch('country_name', '')
         website = org_website(item: item)
+
         # If no website or country then just return the name
-        return item['name'] unless website.present? || country.present?
+        return name unless website.present? || country.present?
 
         # Otherwise return the contextualized name
-        "#{item['name']} (#{website || country})"
+        "#{name} (#{website || country})"
       end
 
-      # Extracts the org's ISO639 if available
+      # Extracts the org's language
+      # {
+      #   "id": "https://ror.org/012345678",
+      #   "names": [
+      #     { "value": "Université de Montréal", "types": ["ror_display"], "lang": "fr" },
+      #     { "value": "University of Montreal", "types": ["alias"], "lang": "en" }
+      #   ]
+      # }
       def org_language(item:)
```

I guess https://github.com/portagenetwork/roadmap/pull/1247/files#r2643715808 also applies to `def org_language(item:)`.

Looking at the duplicate code between def `sort_name(item)` and `def org_language(item:)` maybe you could DRY the two up as well (`def org_name(item:)` might benefit from this as well). Maybe something like the following would work?

```ruby
def parse_results(json:)
  ror_display = ror_display_entry(item) # This will replace name = sort_name(item)
  ...
  name: org_name(item: item, ror_display: ror_display),
  sort_name: sort_name(ror_display),
  ...
  language: org_language(ror_display),
```

```ruby
def ror_display_entry(item)
  item['names'].find { |n| n['types'].include?('ror_display') }
end
```
The other referenced helpers would need small updates as well.

---

I'm thinking a simpler way of doing this would be to just use the ror_display helper in the helper functions instead of parse_results, so it would look something like this 

```ruby
def org_language(item:)
  # ROR uses two-letter / ISO 639-1 language codes (e.g. "en")
  dflt = I18n.default_locale.to_s.split('-').first

  ror_display_entry(item)&.fetch('lang', dflt) || dflt
end
``` 
I think this achieves the DRY goal without overcomplicating `parse_results`.

---

That sounds okay. With that approach, every helper would be re-executing `ror_display` helper. But that might just be a trivial for performance. I think your idea is cool and I like the idea of simplifying parse_results. I have one other suggestion if we go with what you're proposing:
```ruby
name = sort_name(item)
next unless item['id'].present? && name.present?
```
I worry about how the legacy code would just skip to the next item if the condition is false. `item['id'].present? && name.present?` should be guaranteed by the ROR API contract. So if this condition is in fact false, it may signify that something is wrong with our code. Rather than silently iterating to the next item, it could log an error:

```ruby
  json['items'].each do |item|
    begin
      results << {
        ror: item.fetch('id').gsub(/^#{landing_page_url}/, ''),
        name: org_name(item: item),
        sort_name: sort_name(item),
        language: org_language(item),
        url: org_url(item),
        fundref: fundref_id(item: item),
        abbreviation: item.fetch('acronyms', []).first
      }
    rescue KeyError, NoMethodError => e
      # Log error and continue to the next item
      Rails.logger.error("Invalid ROR record: #{e.message}, item: #{item.inspect}")
    end
```

---

That edit to `parse_results` unfortunately doesn't catch the error, it just fills the values with nil. This is what the failing test indicated:

```ruby
it 'ignores items with no name or id' do
        json = { items: [
          { id: Faker::Internet.url, names: [{ types: ['ror_display'], value: Faker::Lorem.word }], external_ids: [] },
          { id: Faker::Internet.url, names: [], external_ids: [] },
          { names: [{ types: ['ror_display'], value: Faker::Lorem.word }], external_ids: [] }
        ] }.to_json

        items = described_class.send(:parse_results, json: JSON.parse(json))
        byebug
        expect(items.length).to eql(1)
      end
```

```
(byebug) items
[{:ror=>"http://block.test/mellisa", :name=>"suscipit", :sort_name=>"ExternalApis::RorService", :url=>nil, :language=>"en", :fundref=>"", :abbreviation=>nil}, {:ror=>"http://torphy.example/mignon", :name=>nil, :sort_name=>"ExternalApis::RorService", :url=>nil, :language=>"en", :fundref=>"", :abbreviation=>nil}]
```

I think keeping the check might be okay if the error logging is added as a safeguard.

---

Hmm, I'm still leaning toward removing that check. Those values are explicitly guaranteed by the ROR API spec. Also, `next unless item['id'].present? && name.present?` doesn't just ignore "bad data", it actively suppresses more useful errors. For instance, this example shows how it could suppress a KeyError:

```
irb(main):001> item = {}
=> {}
irb(main):002> item['id'].present?
=> false
irb(main):003> item.fetch('id').gsub('test', '')
(irb):3:in `fetch': key not found: "id" (KeyError)
        from (irb):3:in `<main>'
irb(main):004>
```

---

Let me know what you think of this.

```ruby
json['items'].each do |item|
      id = item&.fetch('id')
      name = sort_name(item)
      raise KeyError, 'Missing ror_display name' if name.blank?

      results << {
        ror: id.gsub(/^#{landing_page_url}/, ''),
        name: org_name(item: item),
        sort_name: name,
        url: org_url(item),
        language: org_language(item: item),
        fundref: fundref_id(item: item),
        abbreviation: item.fetch('acronyms', []).first
      }
    rescue KeyError, NoMethodError => e
      Rails.logger.error("Invalid ROR record: #{e.message}, item: #{item.inspect}")
    end
 ```
It includes error logging to prevent silent errors while also preventing malinformed items from being passed into `results`.